packages/commerce-sdk-react/src/hooks/ShopperCustomers/cache.ts

```diff
@@ -4,94 +4,253 @@
+                        )
+
+                        // Return undefined (no changes) if no payment instrument was found.
+                        if (
+                            typeof paymentInstrumentIndex === 'undefined' ||
+                            paymentInstrumentIndex < 0
+                        ) {
+                            return
+                        }
+
+                        // Remove the found payment instrument.
+                        customer?.paymentInstruments?.splice(paymentInstrumentIndex, 1)
+
+                        return customer
+                    })
+                }
+            ],
             remove: [{queryKey: getCustomerPaymentInstrument.queryKey(parameters)}]
         }
     },
     deleteCustomerProductList(customerId, {parameters}) {
         return {
-            // TODO: Rather than invalidate, can we selectively update?
-            invalidate: [{queryKey: getCustomerProductLists.queryKey(parameters)}],
+            update: [
+                {
+                    queryKey: getCustomerProductLists.queryKey(parameters),
+                    updater: createUpdateFunction((result: CustomerProductListResult) => {
+                        const listIndex = result.data.findIndex(({id}) => id === parameters.listId)
+
+                        // Return undefined if no list is found
+                        if (listIndex < 0) {
+                            return
+                        }
+
+                        // Remove the list from the result object
+                        result.data.splice(listIndex, 1)
+                        result.limit--
+                        result.total--
+
+                        return result
+                    })
+                }
+            ],
             remove: [{queryKey: getCustomerProductList.path(parameters)}]
         }
     },
     deleteCustomerProductListItem(customerId, {parameters}) {
         return {
-            // TODO: Rather than invalidate, can we selectively update?
-            invalidate: [
-                {queryKey: getCustomerProductList.queryKey(parameters)},
-                {queryKey: getCustomerProductLists.queryKey(parameters)}
+            update: [
+                {
+                    queryKey: getCustomerProductList.queryKey(parameters),
+                    updater: createUpdateFunction((list: CustomerProductList) => {
+                        const itemIndex = list.customerProductListItems?.findIndex(
+                            ({id}) => id === parameters.itemId
+                        )
+
+                        // Return undefined if there is no item found.
+                        if (typeof itemIndex === 'undefined' || itemIndex < 0) {
+                            return
+                        }
+
+                        // Remove the list item
+                        list.customerProductListItems?.splice(itemIndex as number, 1)
```

Comment:
Shouldn't `itemIndex` already be `number` here? If not, try changing the `typeof itemIndex === 'undefined'` to just `itemIndex === undefined`, or maybe add a fallback param `?? -1` so that it's only ever a number.